probe.go

```diff
@@ -960,3 +999,52 @@ func (p *Probe) detachLSM() error {
 	}
 	return nil
 }
+
+// attachPerfEvent - Attaches the perf_event program
+func (p *Probe) attachPerfEvent() error {
+	if p.PerfEventType != unix.PERF_TYPE_HARDWARE && p.PerfEventType != unix.PERF_TYPE_SOFTWARE {
+		return fmt.Errorf("unknown PerfEventType parameter: %v (expected unix.PERF_TYPE_HARDWARE or unix.PERF_TYPE_SOFTWARE)", p.PerfEventType)
+	}
+
+	attr := unix.PerfEventAttr{
+		Type:   uint32(p.PerfEventType),
+		Sample: uint64(p.SamplePeriod),
+		Config: uint64(p.PerfEventConfig),
+	}
+
+	if p.SampleFrequency > 0 {
+		attr.Sample = uint64(p.SampleFrequency)
+		attr.Bits |= unix.PerfBitFreq
+	}
+
+	if p.PerfEventCPUCount == 0 {
+		p.PerfEventCPUCount = runtime.NumCPU()
+	}
+
+	pid := p.PerfEventPID
+	if pid == 0 {
+		pid = -1
+	}
+
+	for cpu := 0; cpu < p.PerfEventCPUCount; cpu++ {
```

Comment:
If you have a partial attachement, ie only part of the perf events are successful and you have a failure in the middle of the loop, should you free those that were successful ?

Reply:
Hey Erwan,
Thanks for the review ! So yes you're 100% right, we need to cleanup successful attempts if the last one failed. The good news is that the `Probe.attach()` method already handles it for us: if `Probe.attachPerfEvent()` returns an error, then `Probe.attach()` will eventually call `_ = p.stop(false)` which will call `Probe.detachPerfEvent()` ... which will free the successful attempts 😃

Reply:
Thanks for pointing that out !